core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/allowances/mareco_impl/AcceleratingSlopeCoast.java

```diff
@@ -1,40 +1,94 @@
-    private double endPos;
+    /** Position where the coasting envelope should merge back into the base envelope */
+    private double endPos = NaN;
 
     /** Average acceleration of the train during the slope */
-    private double slopeAverageAcceleration;
+    private double slopeAverageAcceleration = NaN;
 
     /** An estimate of the average acceleration of the train
      * from the start of coasting to the beginning of the accelerating slope */
-    private double previousAccelerationEstimate;
+    private double previousAccelerationEstimate = NaN;
+
+    /** Position where the train starts accelerating again until it reaches the max speed */
+    private final double accelerationStartPosition;
 
     /** The start and end speed limit of the accelerating slope */
-    private double speedLimit;
+    private final double speedLimit;
+
+    /** Used to compute the mean acceleration over the accelerating slope */
+    private final DistanceAverage meanAccelerationBuilder = new DistanceAverage();
 
     private AcceleratingSlopeCoast(
-            double endPos,
-            double slopeAverageAcceleration,
-            double previousAccelerationEstimate,
+            double accelerationStartPosition,
             double speedLimit
     ) {
-        this.endPos = endPos;
-        this.slopeAverageAcceleration = slopeAverageAcceleration;
-        this.previousAccelerationEstimate = previousAccelerationEstimate;
+        this.accelerationStartPosition = accelerationStartPosition;
         this.speedLimit = speedLimit;
     }
 
+    /** Finish building the slope instance, once we know the end position */
+    private AcceleratingSlopeCoast build(double endPos, EnvelopeSimContext context) {
+        this.slopeAverageAcceleration = meanAccelerationBuilder.getAverage();
+        this.previousAccelerationEstimate = estimatePreviousAcceleration(context);
+        this.endPos = endPos;
+        assert !Double.isNaN(slopeAverageAcceleration);
+        assert !Double.isNaN(endPos);
+        assert !Double.isNaN(previousAccelerationEstimate);
+        assert !Double.isNaN(speedLimit);
+        return this;
+    }
+
+    /** Estimates the average acceleration during the part where the train decelerates */
+    private double estimatePreviousAcceleration(EnvelopeSimContext context) {
+        // We look for the natural acceleration an offset before the start of the slope.
+        // The exact offset is arbitrary, we only need an approximation (as long as there is no discontinuity).
+        // If we assume the train starts accelerating when its center is roughly at the highest point, and
+        // we look for the acceleration on the ramp, half of the rolling stock length is a good offset value.
+        // If we need a better approximation than this, the "next step" would be to average over a certain distance.
+        var accelerationStart = findExactStartAcceleratingSlope(context);
+        var offset = context.rollingStock.getLength() / 2;
```

Comment:
I suggest an offset corresponding to one entire rolling stock length, as you said it's arbitrary but in the case the slope difference is so high that the acceleration spot is not at the middle of the train but rather towards the tail
```suggestion
        var offset = context.rollingStock.getLength();
```